File: qqrobot-spring-boot-starter/src/main/java/de/honoka/qqrobot/starter/component/logger/DefaultRobotLogger.java
package de.honoka.qqrobot.starter.component.logger;

import cn.hutool.core.exceptions.ExceptionUtil;
import de.honoka.qqrobot.framework.api.RobotFramework;
import de.honoka.qqrobot.starter.component.logger.dao.ExceptionRecordDao;
import de.honoka.qqrobot.starter.component.logger.dao.UsageLogDao;
import de.honoka.qqrobot.starter.component.logger.entity.ExceptionRecord;
import de.honoka.qqrobot.starter.component.logger.entity.UsageLog;
import de.honoka.qqrobot.starter.config.ConditionalBeansConfig;
import de.honoka.sdk.spring.starter.core.context.ConditionalComponent;
import de.honoka.sdk.util.various.Retrier;
import jakarta.annotation.Resource;
import org.springframework.context.annotation.Lazy;

import java.util.Date;

@ConditionalComponent(ConditionalBeansConfig.class)
public class DefaultRobotLogger implements RobotLogger {
    
    @Resource
    private UsageLogDao usageLogDao;
    
    @Lazy
    @Resource
    private RobotFramework framework;
    
    @Resource
    private ExceptionRecordDao exceptionRecordDao;
    
    @Override
    public void logException(Throwable t) {
        String str = ExceptionUtil.stacktraceToString(t, -1);
        new Retrier().tryCode(() -> exceptionRecordDao.insert(
            new ExceptionRecord()
                .setDatetime(new Date())
                .setExceptionText(str)
        ));
    }
    
    @Override
    public void logMsgExecution(Long group, long qq, String msg, String reply) {
        new Retrier().tryCode(() -> usageLogDao.insert(
            new UsageLog()
                .setGroupName(framework.getGroupName(group))
                .setQq(qq).setDatetime(new Date())
                .setUsername(framework.getNickOrCard(group, qq))
                .setMsg(msg)
                .setReply(reply)
        ));
    }
}

File: qqrobot-spring-boot-starter/src/main/java/de/honoka/qqrobot/starter/core/logger/DefaultRobotLogger.kt
package de.honoka.qqrobot.starter.core.logger

import cn.hutool.core.exceptions.ExceptionUtil
import de.honoka.qqrobot.framework.api.RobotFramework
import de.honoka.qqrobot.starter.core.logger.dao.ExceptionRecordDao
import de.honoka.qqrobot.starter.core.logger.dao.UsageLogDao
import de.honoka.qqrobot.starter.core.logger.entity.ExceptionRecord
import de.honoka.qqrobot.starter.core.logger.entity.UsageLog
import de.honoka.sdk.util.kotlin.various.tryBlock
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Component
import java.util.*

@ConditionalOnMissingBean(RobotLogger::class)
@Component
class DefaultRobotLogger(
    private val usageLogDao: UsageLogDao,
    @param:Lazy
    private val framework: RobotFramework,
    private val exceptionRecordDao: ExceptionRecordDao
) : RobotLogger {

    override fun logException(t: Throwable) {
        tryBlock(3) {
            val record = ExceptionRecord().apply {
                datetime = Date()
                exceptionText = ExceptionUtil.stacktraceToString(t, -1)
            }
            exceptionRecordDao.insert(record)
        }
    }

    override fun logMsgExecution(group: Long, qq: Long, msg: String, reply: String) {
        tryBlock(3) {
            val log = UsageLog().apply {
                groupName = framework.getGroupName(group)
                this.qq = qq
                datetime = Date()
                username = framework.getNickOrCard(group, qq)
                this.msg = msg
                this.reply = reply
            }
            usageLogDao.insert(log)
        }
    }
}
